feat: implement support for monitoring D-Bus signals

FineFindus opened the pull request:
Implements support for using `dbus-monitor`, which allows direct monitoring of signals, which are send when the settings changes.

This reduces the system load, as continuous polling is no longer required, as well as improving the responsiveness, as the theme is now updated immediately, instead of waiting for the next polling-cycle to complete.

If `dbus-monitor` is not available, the polling method will be used as a fallback.

https://github.com/user-attachments/assets/633cad30-e6cf-4f39-966e-73771dfe2f51

FineFindus committed:
feat: implement support for monitoring D-Bus signals

Implements support for using `dbus-monitor`, which allows direct monitoring of
signals, which are send when the settings changes.

This reduces the system load, as continuous polling is no longer required, as
well as improving the responsiveness, as the theme is now updated immediately,
instead of waiting for the next polling-cycle to complete.

If `dbus-monitor` is not available, the polling method will be used as a
fallback.

aliaksandr-trush commented:
Tested this change and it works great!

With the current implementation (using `dbus-send`) with default opts the plugin could consume 100% CPU (more details could be found in https://github.com/flatpak/xdg-desktop-portal/issues/1416), but with moving to `dbus-monitor` the plugin does not use CPU resources at all.

aliaksandr-trush commented on lua/auto-dark-mode/interval.lua:
```diff
@@ -16,9 +16,9 @@ local uv = vim.uv or vim.loop
-        end
+                if stderr ~= "" then
+                    return nil;
+                end
```
Please format all your changes:
existing files use tabs for indent, but in your changes you are using spaces.

aliaksandr-trush commented on lua/auto-dark-mode/init.lua:
```diff
@@ -54,6 +54,8 @@ M.state = {
 	---@type table
 	query_command = {},
+	---@type string?
+	monitor_command = nil,
```
Could it be table to be similar with query_command type?

FineFindus replied:
Sure, done in c439798ef628b169ac24f830827c00479f95e584. Just looks a bit silly, as it's only three lines.

aliaksandr-trush commented on lua/auto-dark-mode/interval.lua:
```diff
@@ -67,6 +67,33 @@ local function sync_theme(appearance)
+    -- if no callback is provided, use a no-op
+    callback = callback or function() end
+
+    vim.fn.jobstart(M.state.monitor_command,
```
I see in documentation that `vim.system()` is preferable over `vim.fn.jobstart()`. Could it be done in the same way as in other methods?

FineFindus replied:
No, as `vim.system` only provides the command output once the command has finished executing (which is obviously never in the case of the monitoring command). I checked the docs, but there doesn't seem to be the option to parse the output whilst the command is still executing.

aliaksandr-trush replied:
As I see in documentation you can pass function to opts.stdout https://neovim.io/doc/user/lua.html#_lua-module:-vim.system.
Something like this should work:

```
	vim.system(M.state.monitor_command, {
		text = true,
		stdout = function(err, data)
			if string.match(data, "uint32") then
				callback(data, "")
			end
		end,
	})
```

FineFindus replied:
It does, but only if you don't typo the function :facepalm:

aliaksandr-trush reviewed (changes requested):
These changes work good, but code style could be improved.

FineFindus committed:
style: reformat code

FineFindus committed:
refactor: use table for monitor command

FineFindus commented:
This should also solve https://github.com/f-person/auto-dark-mode.nvim/issues/64.

FineFindus committed:
style: format using stylua

FineFindus committed:
refactor: use prefer `vim.system` if available

x1unix commented:
@FineFindus thanks for the solution.

I've found a bug - after neovim instance is closed, the created `dbus-monitor` process still exists.

The fix is here: https://github.com/x1unix/auto-dark-mode.nvim/commit/138a39a26c9419c95f0920e6b4e7bc66b17e043f

For example, I restarted my neovim 3 times and got 2 orphan `dbus-monitor` processes left from previous instances:

<img width="1916" height="412" alt="image" src="https://github.com/user-attachments/assets/47be2ac2-d23e-432e-a6ef-2dc8ae98afbc" />

FineFindus committed:
fix: kill monitor process on exit

Avoids the dbus-monitor process staying alive after exiting, thus leaking the
resource.

Ref: https://github.com/neovim/neovim/issues/29475
Ref: https://github.com/f-person/auto-dark-mode.nvim/pull/63#issuecomment-4130903234

FineFindus commented:
Thanks for the notice and the fix. This seems to be a [Neovim bug](https://github.com/neovim/neovim/issues/29475), which only affects `vim.system`, so I've adjust the fix a bit :)

f-person reviewed (approved):
lgtm!
